ai_challenge/pig_chase/pig_chase_dqn.py

```diff
@@ -0,0 +1,169 @@
+# Copyright (c) 2017 Microsoft Corporation.
+#
+# Permission is hereby granted, free of charge, to any person obtaining a copy of this software and associated
+# documentation files (the "Software"), to deal in the Software without restriction, including without limitation the
+#  rights to use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the Software,
+# and to permit persons to whom the Software is furnished to do so, subject to the following conditions:
+#
+# The above copyright notice and this permission notice shall be included in all copies or substantial portions of
+# the Software.
+#
+# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO
+# THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,
+#  TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+# SOFTWARE.
+# ===================================================================================================================
+
+import os
+import sys
+from argparse import ArgumentParser
+from datetime import datetime
+from os import path
+from threading import Thread, active_count
+from time import sleep
+
+from malmopy.agent import LinearEpsilonGreedyExplorer
+from malmopy.model.chainer import QNeuralNetwork, DQNChain
+
+from common import parse_clients_args, visualize_training, ENV_AGENT_NAMES
+from agent import PigChaseChallengeAgent, PigChaseQLearnerAgent
+from environment import PigChaseEnvironment, PigChaseSymbolicStateBuilder
+from malmopy.environment.malmo import MalmoALEStateBuilder
+
+from malmopy.agent import TemporalMemory
+
+try:
+    from malmopy.visualization.tensorboard import TensorboardVisualizer
+    from malmopy.visualization.tensorboard.cntk import CntkConverter
+except ImportError:
+    print('Cannot import tensorboard, using ConsoleVisualizer.')
+    from malmopy.visualization import ConsoleVisualizer
+
+# Enforce path
+sys.path.insert(0, os.getcwd())
+sys.path.insert(1, os.path.join(os.path.pardir, os.getcwd()))
+
+DQN_FOLDER = 'results/baselines/%s/dqn/%s-%s'
+EPOCH_SIZE = 250000
+
+
+def agent_factory(name, role, clients, device, max_epochs, logdir, visualizer):
+
+    assert len(clients) >= 2, 'Not enough clients (need at least 2)'
+    clients = parse_clients_args(clients)
+
+    if role == 0:
+
+        builder = PigChaseSymbolicStateBuilder()
+        env = PigChaseEnvironment(clients, builder, role=role,
+                                  randomize_positions=True)
+        agent = PigChaseChallengeAgent(name)
+
+        obs = env.reset()
```

Comment:
Needs to switch appearance of the challenge agent based on type. See https://github.com/Microsoft/malmo-challenge/blob/master/ai_challenge/pig_chase/pig_chase_baseline.py#L59